### ISSUE TITLE
Unable to delete a tag within a Tag Group

### ISSUE THREAD

**Issue by Agent-Sesame** [tagspaces/tagspaces#839]
**Description**
I have tags organized into groups. Example: Tag Group called "Data Retention" with 16 tags under it. One tag is called "2018-dec-31-1yr" and since the date has passed, I wish to delete it. This example tag is also the first tag in the tag group in position 1. I click the 3 dot icon on the tag, then select "Delete Tag", then click "Yes" to acknowledge I wish to delete it. Following this process fails to delete the tag in position 1 but instead deletes the tag in the end position of the tag group list. I've been accidentally deleting tags due to this bug.

**To Reproduce**
1. Go to 'Tag Group Operation'
2. Click on > to expand tag group, example "Data Retention"
3. Right click on the first tag in the group, example "2018-dec-31-1yr"
4. In the pop up menu, select "Delete Tag"
5. In the next confirmation menu, select "Yes" to confirm the deletion request.
6. Instead of tag "2018-dec-31-1yr" being deleted, the last tag in the list is deleted, example "transient"

**Expected behavior**
I expect the tag I right clicked on to be deleted, not the last tag in the tag group.

**Screenshots**
Does not seem demonstrable with screen prints.

**Desktop (please complete the following information):**
 - OS: MacOS High Sierra v10.13.6
 - Browser: N/A
 - TagSpaces Version 3.0.1

**Smartphone (please complete the following information):**
N/A

**Additional context**
N/A

**Comment by Agent-Sesame** [comment]
@uggrock is version 3.0.3 released?

**Comment by uggrock** [comment]
Yes it was released as pre-release, today we released 3.0.4, you can find it [here](https://github.com/tagspaces/tagspaces/releases)

**Comment by Agent-Sesame** [comment]
@uggrock Hi, downloaded 3.0.4 version here and this error behavior persists. I've screen shotted a number of actions. Compare the tag list in picture #1 against picture #5. I tried to delete the first gray tag, see picture #3 to check I've selected tag '2018-dec-31-1yr'. What is deleted is tag '2024-dec-31-7yr'. Picture #5 shows that tag now missing in the list. 

![1 beginning state](https://user-images.githubusercontent.com/8811189/51150903-5a52ba00-181d-11e9-9b61-1a406f96d075.png)
![2 right click top gray tag](https://user-images.githubusercontent.com/8811189/51150904-5a52ba00-181d-11e9-9443-c3dbf1d3ca19.png)
![3 confirmation message](https://user-images.githubusercontent.com/8811189/51150905-5a52ba00-181d-11e9-8eb8-3d935b57ab68.png)
![4 selecting yes](https://user-images.githubusercontent.com/8811189/51150906-5a52ba00-181d-11e9-981d-b4312425036b.png)
![5 final state with deletion erro](https://user-images.githubusercontent.com/8811189/51150907-5a52ba00-181d-11e9-965d-23530cac52f0.png)

**Comment by Agent-Sesame** [comment]
Is the release announced Jan 18, 2019 via email the same as the prereleased v3.0.4 you suggested I test?

**Comment by uggrock** [comment]
There are some differences but not in the area related to your issue. Is the issue for one particular tag, or for any tag in any tag group? If it is just for one tag, which will be very strange, you can try to delete the its tag group.

**Comment by Agent-Sesame** [comment]
Updated to the Jan 18th announced download, version TS Pro 3.0.5. I ran three tests, 2 failed and 2 passed. Given these results, what do you suggest?

Test 1 - FAIL - In the same tag group included the screen shots above, I tried to delete a different tag "2019-dec-31-1yr". What was deleted was "2023-dec-31-7yr". Note this tag group and tags were custom additions and imported from prior TS Pro json file. Screen print evidence below. 

Test 2 - FAIL - In a different tag group "Common Tags", this tag group was default native to the original TS Pro installation, I tried to delete tag "pcso". What was deleted was tag "movies". While this tag group was native to TS Pro, I added my own custom tags. Customizations were imported from prior TS Pro json file. See screen print evidence below.

Test 3 - PASS - Created new tag group "test tag group". Created tags "tag_1", "tag_2", "tag_3". Deleting any of these tags deleted the intended tag.

Test 4 - PASS - In same tag group as Test 3, Created tags "longtagnameLongTagName1", "longtagnameLongTagName2", "longtagnameLongTagName3". Deleting any of these tags deleted the intended tag. 


![image](https://user-images.githubusercontent.com/8811189/51500204-1b26ea80-1d82-11e9-9d47-88e38b29cfbf.png)

![image](https://user-images.githubusercontent.com/8811189/51500219-24b05280-1d82-11e9-87f7-471561fcc7bf.png)

![image](https://user-images.githubusercontent.com/8811189/51500233-2f6ae780-1d82-11e9-9e96-333aa404b0ef.png)

![image](https://user-images.githubusercontent.com/8811189/51500243-38f44f80-1d82-11e9-9c27-e3412174a567.png)

**Comment by Agent-Sesame** [comment]
Does the Mar 1, 2019 release fix this issue?

**Comment by rlach** [comment]
I also have issues deleting tags which were added to tagspaces from import/export procedure from older version. Version 3.1.14. I'm adding tag library export which breaks the deletion.

[tag-library-broken.zip](https://github.com/tagspaces/tagspaces/files/3204194/tag-library-broken.zip)

Try importing this. Then from `Ratings` tag group try removing `+5`. It will instead remove `+9`. I'm convinced the reason for this behavior is the fact the old tags do not have `id` field set. I've even wrote small script which added uuids to all tags in my export.json and tags work properly now.

**Comment by iliansap** [comment]
The problem for this is that for some reason your tag does not have id property. If you add unique id and then import the file everything should work fine.
                {
                    "id": "3129b718-5dbf-4227-a1db-d55bc77bc86e",
                    "type": "plain",
                    "title": "+7",
                    "color": "#ffcc24",
                    "textcolor": "#ffffff",
                    "description": "+7",
                    "icon": "",
                    "style": "color: #ffffff !important; background-color: #ffcc24 !important;"
                },

**Comment by Agent-Sesame** [comment]
@iliansap could you describe how to add an ID? I don't recall any particular feature in the tag library GUI that permits ID value modification.

**Comment by iliansap** [comment]
You can just create the tags in tagspaces and export them. The tags will be created with ID automatically.

**Comment by rlach** [comment]
Tagspaces2 did not add ids to the export. Tagspaces3 is broken since it doesn't add ids on import when they are missing. When tags from tagspaces2 are added to tagpsaces3 without ids tagspaces3 will never add them anymore, even during export.

In my opinion this is a bug in tagspaces3. Tagspaces3 should either fix or reject wrong data during import. Currently it doesn't, so when someone implements a fix for the importer they should also make a check that runs on startup that confirms that all tags have uuids, and if not add them(should be necessary only once).

Users shouldn't be required to manually recreate all tags and reorganize them into groups because of a bug in the application. I have about 400 tags, each tag group has different coloring, I would spend a day reorganizing them manually, especially since currently we cannot drag and drop tags between groups so even if I gathered all tags from all my folders by using the premium feature I would still have mess in my tags.

**Comment by Agent-Sesame** [comment]
thanks @rlach I think you've technically described my problem better than I was able to. I purchased a pro version of TagSpaces Nov 2017 so I suspect that initial purchase was a v2 purchase. I definitely recall importing / exporting the tag scheme at least once during an upgrade.

I tried recreating new tags and groups as per @iliansap suggestion a couple of days ago and it turned into a giant jumblefark and gave up on the task. Regrettably, I have v2 legacy tags intermingled in a single group with v3 new tags, so I can't tell which tags are faulty and which are not.
